crates/server/src/admin/handlers/blob.rs

```diff
@@ -161,28 +285,252 @@ pub async fn info_handler(
+        Ok(Some(blob_meta)) => {
+            let content_type = detect_mime_type(&state, &blob_id)
+                .await
+                .unwrap_or_else(|| "application/octet-stream".to_string());
+            let etag = format!("\"{}\"", hex::encode(blob_meta.hash));
+
+            Response::builder()
+                .status(StatusCode::OK)
+                .header("Content-Length", blob_meta.size.to_string())
+                .header("Content-Type", &content_type)
+                .header("ETag", &etag)
+                .header("Cache-Control", "public, max-age=3600") // 1 hour cache
+                .header("X-Blob-ID", blob_id.to_string())
+                .header("X-Blob-Hash", hex::encode(blob_meta.hash))
+                .header("X-Blob-MIME-Type", &content_type)
+                .header(
+                    "Access-Control-Expose-Headers",
+                    "X-Blob-ID, X-Blob-Hash, X-Blob-MIME-Type, ETag",
+                )
+                .body(Body::empty())
+                .unwrap_or_else(|_| {
+                    ApiError {
+                        status_code: StatusCode::INTERNAL_SERVER_ERROR,
+                        message: "Failed to build response".to_owned(),
+                    }
+                    .into_response()
+                })
+        }
         Ok(None) => ApiError {
             status_code: StatusCode::NOT_FOUND,
             message: "Blob not found".to_owned(),
         }
         .into_response(),
         Err(err) => {
-            tracing::error!("Failed to check blob {}: {:?}", blob_id, err);
+            tracing::error!("Failed to get blob metadata: {:?}", err);
             ApiError {
                 status_code: StatusCode::INTERNAL_SERVER_ERROR,
-                message: format!("Failed to check blob: {}", err),
+                message: format!("Failed to retrieve blob metadata: {}", err),
             }
             .into_response()
         }
     }
 }
 
-pub use upload_handler as handler;
+/// Detect MIME type by reading the first few bytes of a blob
+async fn detect_mime_type(state: &AdminState, blob_id: &BlobId) -> Option<String> {
+    match state.node_client.get_blob(blob_id) {
+        Ok(Some(mut blob_stream)) => {
+            use futures_util::StreamExt;
+
+            if let Some(Ok(first_chunk)) = blob_stream.next().await {
+                let bytes = first_chunk.as_ref();
+                let sample_size = std::cmp::min(bytes.len(), 20);
+                return Some(detect_mime_from_bytes(&bytes[..sample_size]));
+            }
+        }
+        Ok(None) => {
+            tracing::warn!("Blob {} not found for MIME detection", blob_id);
+        }
+        Err(err) => {
+            tracing::warn!(
+                "Failed to read blob {} for MIME detection: {:?}",
+                blob_id,
+                err
+            );
+        }
+    }
+
+    None
+}
+
+/// Detect MIME type from file magic bytes
+fn detect_mime_from_bytes(bytes: &[u8]) -> String {
```

Comment:
https://docs.rs/infer would've been nice, but for a first impl this is tolerable